src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/Whitelist.java

```diff
@@ -66,6 +69,66 @@
 
     public abstract boolean permitsStaticFieldSet(@Nonnull Field field, @CheckForNull Object value);
 
+    // Utility methods for creating canonical string representations of the signature
+    public static final StringBuilder joinWithSpaces(StringBuilder b, String[] types) {
+        for (String type : types) {
+            b.append(' ').append(type);
+        }
+        return b;
+    }
+
+    @Restricted(NoExternalUse.class)
+    public static String[] argumentTypes(Class<?>[] argumentTypes) {
+        String[] s = new String[argumentTypes.length];
+        for (int i = 0; i < argumentTypes.length; i++) {
+            s[i] = EnumeratingWhitelist.getName(argumentTypes[i]);
+        }
+        return s;
+    }
+
+    /** Canonical name for a field access. */
+    @Restricted(NoExternalUse.class)
+    public static String canonicalFieldString(@Nonnull Field field) {
+        return EnumeratingWhitelist.getName(field.getDeclaringClass()) + ' ' + field.getName();
+    }
+
+    /** Canonical name for a method call. */
+    @Restricted(NoExternalUse.class)
+    public static String canonicalMethodString(@Nonnull Method method) {
+        return joinWithSpaces(new StringBuilder(EnumeratingWhitelist.getName(method.getDeclaringClass())).append(' ').append(method.getName()), argumentTypes(method.getParameterTypes())).toString();
+    }
+
+    /** Canonical name for a constructor call. */
+    @Restricted(NoExternalUse.class)
+    public static String canonicalConstructorString(@Nonnull Constructor cons) {
+        return joinWithSpaces(new StringBuilder(EnumeratingWhitelist.getName(cons.getDeclaringClass())), argumentTypes(cons.getParameterTypes())).toString();
+    }
+
+    @Restricted(NoExternalUse.class)
+    public static String canonicalMethodSig(@Nonnull Method method) {
```

Comment:
All of the `canonicalXSig` methods need to generate results matching the output of `toString` for the corresponding `EnumeratingWhiteList.Signature` implementation.  Verified by test: https://github.com/jenkinsci/script-security-plugin/pull/180/files#diff-4ff0624894ffcd88f5f79bcb89bbc036R104

Reply:
Because these match the Signature toStrings, we can use either form as a key -- and can easily generate a canonical String for a permission check based on the Field/Method/etc. 

Note: we could go and rip out a lot of the xxSignature methods etc except that they're published interfaces.